File: cpp/test_timehash.cpp
/*
 * Timehash
 * Copyright (c) 2026-present NAVER Corp.
 * Apache-2.0
 */

#include <iostream>
#include <vector>
#include <string>
#include <cassert>
#include <algorithm>
#include "timehash.hpp"

// Helper function to check if vector contains a value
bool contains(const std::vector<std::string>& vec, const std::string& value) {
    return std::find(vec.begin(), vec.end(), value) != vec.end();
}

// Helper function to check if any query term matches index terms
bool hasMatch(const std::vector<std::string>& queryTerms, const std::vector<std::string>& indexTerms) {
    for (const auto& qk : queryTerms) {
        if (contains(indexTerms, qk)) {
            return true;
        }
    }
    return false;
}

void testGetQueryTerms() {
    std::cout << "Testing getQueryTerms for 14:30...\n";
    Timehash th;
    std::vector<std::string> terms = th.getQueryTerms(1430);
    
    assert(terms.size() == 5);
    assert(terms[0] == "12");
    assert(terms[1] == "1214");
    assert(terms[2] == "121430");
    assert(terms[3] == "12143030");
    assert(terms[4] == "1214303030");
    
    std::cout << "  ✓ testGetQueryTerms passed\n";
}

void testGetQueryTermsInvalid() {
    std::cout << "Testing getQueryTerms with invalid input...\n";
    Timehash th;
    std::vector<std::string> terms = th.getQueryTerms(10001);
    
    assert(terms.size() == 0);
    
    std::cout << "  ✓ testGetQueryTermsInvalid passed\n";
}

void testGetIndexTermsSimple() {
    std::cout << "Testing getIndexTerms for simple range (12:00-13:00)...\n";
    Timehash th;
    std::vector<std::string> terms = th.getIndexTerms("1200", "1300");
    
    assert(contains(terms, "1212"));
    
    std::cout << "  ✓ testGetIndexTermsSimple passed\n";
}

void testGetIndexTermsComplex() {
    std::cout << "Testing getIndexTerms for complex range (11:40-21:00)...\n";
    Timehash th;
    std::vector<std::string> terms = th.getIndexTerms("1140", "2100");
    
    assert(contains(terms, "08113040"));
    assert(contains(terms, "081145"));
    assert(contains(terms, "12"));
    assert(contains(terms, "16"));
    assert(contains(terms, "2020"));
    
    std::cout << "  ✓ testGetIndexTermsComplex passed\n";
}

void testQueryMatchesIndex() {
    std::cout << "Testing query matches index...\n";
    Timehash th;
    
    // Business: 11:40 - 21:00
    std::vector<std::string> indexTerms = th.getIndexTerms("1140", "2100");
    
    // Query: 14:30 (should match)
    std::vector<std::string> queryTerms = th.getQueryTerms(1430);
    assert(hasMatch(queryTerms, indexTerms));
    
    // Query: 10:00 (should not match)
    std::vector<std::string> queryTerms10 = th.getQueryTerms(1000);
    assert(!hasMatch(queryTerms10, indexTerms));
    
    std::cout << "  ✓ testQueryMatchesIndex passed\n";
}

int main() {
    std::cout << "Running Timehash C++ tests...\n\n";
    
    try {
        testGetQueryTerms();
        testGetQueryTermsInvalid();
        testGetIndexTermsSimple();
        testGetIndexTermsComplex();
        testQueryMatchesIndex();
        
        std::cout << "\nAll tests passed! ✓\n";
        return 0;
    } catch (const std::exception& e) {
        std::cerr << "Test failed with exception: " << e.what() << "\n";
        return 1;
    } catch (...) {
        std::cerr << "Test failed with unknown exception\n";
        return 1;
    }
}

File: python/tests/test_timehash.py
# Timehash
# Copyright (c) 2026-present NAVER Corp.
# Apache-2.0

import unittest
from timehash import Timehash


class TestTimehash(unittest.TestCase):
    def setUp(self):
        self.th = Timehash()

    def test_get_query_terms(self):
        """Test query term generation for a specific time."""
        terms = self.th.get_query_terms(1430)  # 14:30
        self.assertEqual(len(terms), 5)
        self.assertEqual(terms[0], "12")
        self.assertEqual(terms[1], "1214")
        self.assertEqual(terms[2], "121430")
        # Level 4 (5-minute): 14:30 is in the 5-minute block starting at 14:30, so key is 30
        self.assertEqual(terms[3], "12143030")
        # Level 5 (1-minute): 14:30 is exactly at minute 30, so key is 30
        self.assertEqual(terms[4], "1214303030")

    def test_get_query_terms_invalid(self):
        """Test query term generation with invalid input."""
        # 10000 is treated as 100:00 which is invalid, but code generates keys
        # C++ code checks hhmm > 10000, so 10000 itself is processed
        terms = self.th.get_query_terms(10001)  # Actually invalid (> 10000)
        self.assertEqual(len(terms), 0)

    def test_get_index_terms_simple(self):
        """Test index term generation for a simple range."""
        terms = self.th.get_index_terms("1200", "1300")  # 12:00 - 13:00
        self.assertIn("1212", terms)

    def test_get_index_terms_complex(self):
        """Test index term generation for a complex range."""
        terms = self.th.get_index_terms("1140", "2100")  # 11:40 - 21:00
        expected_keys = ["08113040", "081145", "12", "16", "2020"]
        for key in expected_keys:
            self.assertIn(key, terms)

    def test_custom_measures(self):
        """Test with custom measures."""
        custom_th = Timehash(measures=[240, 60, 15, 5, 1])
        terms = custom_th.get_index_terms("1200", "1300")
        self.assertGreater(len(terms), 0)

    def test_query_matches_index(self):
        """Test that query keys match index keys correctly."""
        # Business: 11:40 - 21:00
        index_terms = self.th.get_index_terms("1140", "2100")
        
        # Query: 14:30 (should match)
        query_terms = self.th.get_query_terms(1430)
        matches = [q for q in query_terms if q in index_terms]
        self.assertGreater(len(matches), 0, "Query at 14:30 should match business open 11:40-21:00")
        
        # Query: 10:00 (should not match)
        query_terms_10 = self.th.get_query_terms(1000)
        matches_10 = [q for q in query_terms_10 if q in index_terms]
        self.assertEqual(len(matches_10), 0, "Query at 10:00 should not match business open 11:40-21:00")


if __name__ == '__main__':
    unittest.main()
